File: src/potatoclient/java/WindowEventHandler.java
package potatoclient.java;

import javax.swing.*;
import java.awt.*;
import java.awt.event.*;
import java.util.Map;
import java.util.HashMap;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicLong;
import java.util.function.Consumer;
import static potatoclient.java.Constants.*;

public class WindowEventHandler {
    public interface EventCallback {
        void onWindowEvent(EventFilter.EventType type, String eventName, Map<String, Object> details);
        void onWindowClosing();
    }
    
    private final JFrame frame;
    private final EventCallback callback;
    private final EventFilter eventFilter;
    private final ScheduledExecutorService eventThrottleExecutor;
    
    // Window state tracking
    private volatile Point lastWindowLocation;
    private volatile Dimension lastWindowSize;
    private volatile int lastWindowState = JFrame.NORMAL;
    private final AtomicLong lastWindowEventTime = new AtomicLong(0);
    
    // Throttled event tasks
    private volatile ScheduledFuture<?> pendingWindowResizeTask;
    private volatile ScheduledFuture<?> pendingWindowMoveTask;
    private volatile ComponentEvent lastResizeEvent;
    private volatile ComponentEvent lastMoveEvent;
    
    public WindowEventHandler(JFrame frame, EventCallback callback, 
                            EventFilter eventFilter, ScheduledExecutorService eventThrottleExecutor) {
        this.frame = frame;
        this.callback = callback;
        this.eventFilter = eventFilter;
        this.eventThrottleExecutor = eventThrottleExecutor;
        
        // Initialize tracking variables
        this.lastWindowLocation = frame.getLocation();
        this.lastWindowSize = frame.getSize();
    }
    
    public void attachListeners() {
        // Add comprehensive window listeners
        frame.addWindowListener(new WindowAdapter() {
            @Override
            public void windowClosing(WindowEvent e) {
                sendFilteredWindowEvent(EventFilter.EventType.WINDOW_CLOSING, "closing", null);
                callback.onWindowClosing();
            }
            
            @Override
            public void windowOpened(WindowEvent e) {
                sendFilteredWindowEvent(EventFilter.EventType.WINDOW_OPENED, "opened", null);
            }
            
            @Override
            public void windowIconified(WindowEvent e) {
                sendFilteredWindowEvent(EventFilter.EventType.WINDOW_MINIMIZED, "minimized", null);
            }
            
            @Override
            public void windowDeiconified(WindowEvent e) {
                sendFilteredWindowEvent(EventFilter.EventType.WINDOW_RESTORED, "restored", null);
            }
            
            @Override
            public void windowActivated(WindowEvent e) {
                sendFilteredWindowEvent(EventFilter.EventType.WINDOW_FOCUSED, "focused", null);
            }
            
            @Override
            public void windowDeactivated(WindowEvent e) {
                sendFilteredWindowEvent(EventFilter.EventType.WINDOW_UNFOCUSED, "unfocused", null);
            }
        });
        
        // Window state listener for maximize/normal
        frame.addWindowStateListener(e -> {
            int oldState = e.getOldState();
            int newState = e.getNewState();
            
            if (oldState != newState) {
                Map<String, Object> details = new HashMap<>();
                details.put("oldState", getStateString(oldState));
                details.put("newState", getStateString(newState));
                
                if ((newState & Frame.MAXIMIZED_BOTH) == Frame.MAXIMIZED_BOTH) {
                    sendFilteredWindowEvent(EventFilter.EventType.WINDOW_MAXIMIZED, "maximized", details);
                } else if ((oldState & Frame.MAXIMIZED_BOTH) == Frame.MAXIMIZED_BOTH && 
                           (newState & Frame.MAXIMIZED_BOTH) == 0) {
                    sendFilteredWindowEvent(EventFilter.EventType.WINDOW_UNMAXIMIZED, "unmaximized", details);
                }
                
                lastWindowState = newState;
            }
        });
        
        // Component listener for resize and move
        frame.addComponentListener(new ComponentAdapter() {
            @Override
            public void componentResized(ComponentEvent e) {
                lastResizeEvent = e;
                scheduleWindowEvent("resized", () -> {
                    if (lastResizeEvent != null && frame != null) {
                        Dimension newSize = frame.getSize();
                        if (lastWindowSize == null || !newSize.equals(lastWindowSize)) {
                            Map<String, Object> details = new HashMap<>();
                            details.put("width", newSize.width);
                            details.put("height", newSize.height);
                            if (lastWindowSize != null) {
                                details.put("oldWidth", lastWindowSize.width);
                                details.put("oldHeight", lastWindowSize.height);
                            }
                            sendFilteredWindowEvent(EventFilter.EventType.WINDOW_RESIZED, "resized", details);
                            lastWindowSize = newSize;
                        }
                    }
                }, pendingWindowResizeTask, task -> pendingWindowResizeTask = task);
            }
            
            @Override
            public void componentMoved(ComponentEvent e) {
                lastMoveEvent = e;
                scheduleWindowEvent("moved", () -> {
                    if (lastMoveEvent != null && frame != null) {
                        Point newLocation = frame.getLocation();
                        if (lastWindowLocation == null || !newLocation.equals(lastWindowLocation)) {
                            Map<String, Object> details = new HashMap<>();
                            details.put("x", newLocation.x);
                            details.put("y", newLocation.y);
                            if (lastWindowLocation != null) {
                                details.put("oldX", lastWindowLocation.x);
                                details.put("oldY", lastWindowLocation.y);
                            }
                            sendFilteredWindowEvent(EventFilter.EventType.WINDOW_MOVED, "moved", details);
                            lastWindowLocation = newLocation;
                        }
                    }
                }, pendingWindowMoveTask, task -> pendingWindowMoveTask = task);
            }
        });
    }
    
    private void scheduleWindowEvent(String eventType, Runnable task, ScheduledFuture<?> currentTask, 
                                   Consumer<ScheduledFuture<?>> taskSetter) {
        long now = System.currentTimeMillis();
        long lastTime = lastWindowEventTime.get();
        
        if (now - lastTime >= WINDOW_EVENT_THROTTLE_MS) {
            if (lastWindowEventTime.compareAndSet(lastTime, now)) {
                task.run();
                // Cancel any pending task
                if (currentTask != null) {
                    currentTask.cancel(false);
                }
            }
        } else {
            // Cancel existing task and schedule new one
            if (currentTask != null) {
                currentTask.cancel(false);
            }
            
            long delay = WINDOW_EVENT_THROTTLE_MS - (now - lastTime);
            var newTask = eventThrottleExecutor.schedule(() -> {
                task.run();
                lastWindowEventTime.set(System.currentTimeMillis());
            }, delay, TimeUnit.MILLISECONDS);
            taskSetter.accept(newTask);
        }
    }
    
    private String getStateString(int state) {
        var sb = new StringBuilder(32);
        if ((state & Frame.NORMAL) == Frame.NORMAL || state == 0) sb.append("normal,");
        if ((state & Frame.ICONIFIED) == Frame.ICONIFIED) sb.append("minimized,");
        if ((state & Frame.MAXIMIZED_HORIZ) == Frame.MAXIMIZED_HORIZ) sb.append("maximized_horizontal,");
        if ((state & Frame.MAXIMIZED_VERT) == Frame.MAXIMIZED_VERT) sb.append("maximized_vertical,");
        if ((state & Frame.MAXIMIZED_BOTH) == Frame.MAXIMIZED_BOTH) sb.append("maximized,");
        return sb.length() > 0 ? sb.substring(0, sb.length() - 1) : "";
    }
    
    private void sendFilteredWindowEvent(EventFilter.EventType type, String eventName, Map<String, Object> details) {
        if (!eventFilter.isFiltered(type)) {
            callback.onWindowEvent(type, eventName, details);
        }
    }
    
    public void cleanup() {
        if (pendingWindowResizeTask != null) {
            pendingWindowResizeTask.cancel(false);
        }
        if (pendingWindowMoveTask != null) {
            pendingWindowMoveTask.cancel(false);
        }
    }
    
    public Point getLastWindowLocation() {
        return lastWindowLocation;
    }
    
    public Dimension getLastWindowSize() {
        return lastWindowSize;
    }
    
    public int getLastWindowState() {
        return lastWindowState;
    }
}

File: src/potatoclient/kotlin/WindowEventHandler.kt
package potatoclient.kotlin

import java.awt.Dimension
import java.awt.Frame
import java.awt.Point
import java.awt.event.ComponentAdapter
import java.awt.event.ComponentEvent
import java.awt.event.WindowAdapter
import java.awt.event.WindowEvent
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicLong
import javax.swing.JFrame

class WindowEventHandler(
    private val frame: JFrame,
    private val callback: EventCallback,
    private val eventFilter: EventFilter,
    private val eventThrottleExecutor: ScheduledExecutorService,
) {
    interface EventCallback {
        fun onWindowEvent(
            type: EventFilter.EventType,
            eventName: String,
            details: Map<String, Any>?,
        )
    }

    // Window state tracking
    @Volatile private var lastWindowLocation: Point? = frame.location

    @Volatile private var lastWindowSize: Dimension? = frame.size
    private val lastWindowEventTime = AtomicLong(0)

    // Throttled event tasks
    @Volatile private var pendingWindowResizeTask: ScheduledFuture<*>? = null

    @Volatile private var pendingWindowMoveTask: ScheduledFuture<*>? = null

    @Volatile private var lastResizeEvent: ComponentEvent? = null

    @Volatile private var lastMoveEvent: ComponentEvent? = null

    fun attachListeners() {
        // Add comprehensive window listeners
        frame.addWindowListener(
            object : WindowAdapter() {
                // Removed windowClosing - handled by FrameManager

                override fun windowOpened(e: WindowEvent) {
                    sendFilteredWindowEvent(EventFilter.EventType.WINDOW_OPENED, "opened", null)
                }

                override fun windowIconified(e: WindowEvent) {
                    sendFilteredWindowEvent(EventFilter.EventType.WINDOW_MINIMIZED, "minimized", null)
                }

                override fun windowDeiconified(e: WindowEvent) {
                    sendFilteredWindowEvent(EventFilter.EventType.WINDOW_RESTORED, "restored", null)
                }

                override fun windowActivated(e: WindowEvent) {
                    sendFilteredWindowEvent(EventFilter.EventType.WINDOW_FOCUSED, "focused", null)
                }

                override fun windowDeactivated(e: WindowEvent) {
                    sendFilteredWindowEvent(EventFilter.EventType.WINDOW_UNFOCUSED, "unfocused", null)
                }
            },
        )

        // Window state listener for maximize/normal
        frame.addWindowStateListener { e ->
            val oldState = e.oldState
            val newState = e.newState

            if (oldState != newState) {
                val details =
                    mutableMapOf<String, Any>(
                        "oldState" to getStateString(oldState),
                        "newState" to getStateString(newState),
                    )

                if ((newState and Frame.MAXIMIZED_BOTH) == Frame.MAXIMIZED_BOTH) {
                    sendFilteredWindowEvent(EventFilter.EventType.WINDOW_MAXIMIZED, "maximized", details)
                } else if ((oldState and Frame.MAXIMIZED_BOTH) == Frame.MAXIMIZED_BOTH &&
                    (newState and Frame.MAXIMIZED_BOTH) == 0
                ) {
                    sendFilteredWindowEvent(EventFilter.EventType.WINDOW_UNMAXIMIZED, "unmaximized", details)
                }
            }
        }

        // Component listener for resize and move
        frame.addComponentListener(
            object : ComponentAdapter() {
                override fun componentResized(e: ComponentEvent) {
                    lastResizeEvent = e
                    scheduleWindowEvent(
                        task = {
                            lastResizeEvent?.let {
                                val newSize = frame.size
                                if (newSize != lastWindowSize) {
                                    val details =
                                        mutableMapOf<String, Any>(
                                            "width" to newSize.width,
                                            "height" to newSize.height,
                                        )
                                    lastWindowSize?.let {
                                        details["oldWidth"] = it.width
                                        details["oldHeight"] = it.height
                                    }
                                    sendFilteredWindowEvent(EventFilter.EventType.WINDOW_RESIZED, "resized", details)
                                    lastWindowSize = newSize
                                }
                            }
                        },
                        currentTask = pendingWindowResizeTask,
                        taskSetter = { pendingWindowResizeTask = it },
                    )
                }

                override fun componentMoved(e: ComponentEvent) {
                    lastMoveEvent = e
                    scheduleWindowEvent(
                        task = {
                            lastMoveEvent?.let {
                                val newLocation = frame.location
                                if (newLocation != lastWindowLocation) {
                                    val details =
                                        mutableMapOf<String, Any>(
                                            "x" to newLocation.x,
                                            "y" to newLocation.y,
                                        )
                                    lastWindowLocation?.let {
                                        details["oldX"] = it.x
                                        details["oldY"] = it.y
                                    }
                                    sendFilteredWindowEvent(EventFilter.EventType.WINDOW_MOVED, "moved", details)
                                    lastWindowLocation = newLocation
                                }
                            }
                        },
                        currentTask = pendingWindowMoveTask,
                        taskSetter = { pendingWindowMoveTask = it },
                    )
                }
            },
        )
    }

    private fun scheduleWindowEvent(
        task: () -> Unit,
        currentTask: ScheduledFuture<*>?,
        taskSetter: (ScheduledFuture<*>?) -> Unit,
    ) {
        val now = System.currentTimeMillis()
        val lastTime = lastWindowEventTime.get()

        if (now - lastTime >= Constants.WINDOW_EVENT_THROTTLE_MS) {
            if (lastWindowEventTime.compareAndSet(lastTime, now)) {
                task()
                // Cancel any pending task
                currentTask?.cancel(false)
            }
        } else {
            // Cancel existing task and schedule new one
            currentTask?.cancel(false)

            val delay = Constants.WINDOW_EVENT_THROTTLE_MS - (now - lastTime)
            val newTask =
                eventThrottleExecutor.schedule({
                    task()
                    lastWindowEventTime.set(System.currentTimeMillis())
                }, delay, TimeUnit.MILLISECONDS)
            taskSetter(newTask)
        }
    }

    private fun getStateString(state: Int): String =
        buildString {
            append("normal,")
            if ((state and Frame.ICONIFIED) == Frame.ICONIFIED) append("minimized,")
            if ((state and Frame.MAXIMIZED_HORIZ) == Frame.MAXIMIZED_HORIZ) append("maximized_horizontal,")
            if ((state and Frame.MAXIMIZED_VERT) == Frame.MAXIMIZED_VERT) append("maximized_vertical,")
            if ((state and Frame.MAXIMIZED_BOTH) == Frame.MAXIMIZED_BOTH) append("maximized,")
        }.removeSuffix(",")

    private fun sendFilteredWindowEvent(
        type: EventFilter.EventType,
        eventName: String,
        details: Map<String, Any>?,
    ) {
        if (eventFilter.isUnfiltered(type)) {
            callback.onWindowEvent(type, eventName, details)
        }
    }

    fun cleanup() {
        pendingWindowResizeTask?.cancel(false)
        pendingWindowMoveTask?.cancel(false)
    }
}
